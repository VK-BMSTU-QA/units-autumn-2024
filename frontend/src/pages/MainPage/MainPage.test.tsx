import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MainPage } from './MainPage';
import { useProducts } from '../../hooks';
import { Product } from '../../types';

const products: Product[] = [
    {
        name: 'Iphone',
        description: 'Iphone_desc',
        price: 999,
        priceSymbol: '$',
        category: 'Электроника',
        imgUrl: 'imageUrl_1',
        id: 1,
    },
    {
        name: 'lamp',
        description: 'lamp_desc',
        price: 150,
        priceSymbol: '$',
        category: 'Для дома',
        imgUrl: 'imageUrl',
        id: 2,
    },
];

jest.mock('../../hooks', () => ({
    ...jest.requireActual('../../hooks'),
    useProducts: jest.fn(),
}));

describe('MainPage test', () => {
    let setIntervalSpy: jest.SpyInstance;
    let clearIntervalSpy: jest.SpyInstance;

    beforeEach(() => {
        (useProducts as jest.Mock).mockReturnValue(products);

        jest.useFakeTimers();
        setIntervalSpy = jest.spyOn(global, 'setInterval');
        clearIntervalSpy = jest.spyOn(global, 'clearInterval');
        jest.spyOn(global, 'Date').mockImplementation(
            () =>
                ({
                    toLocaleTimeString: () => '12:00:00',
                } as unknown as Date)
        );
    });

    afterEach(() => {
        jest.clearAllTimers();
        setIntervalSpy.mockRestore();
        clearIntervalSpy.mockRestore();
        (global.Date as unknown as jest.Mock).mockRestore();
    });



    it('should clear the interval on unmount', () => {
        const { unmount } = render(<MainPage />);
        expect(setIntervalSpy).toHaveBeenCalledTimes(1);

        unmount();
        expect(clearIntervalSpy).toHaveBeenCalledTimes(1);
    });

    it('should render correctly', () => {
        const rendered = render(<MainPage />);

        expect(rendered.asFragment()).toMatchSnapshot();
    });

    it('should call callback when category click', () => {
        const rendered = render(<MainPage />);
        fireEvent.click(rendered.getAllByText('Для дома')[0]);
        expect(rendered.getByText('lamp')).toBeInTheDocument();
    });

    it('should render ProductCard for everything after unselecting all categories', () => {
        const rendered = render(<MainPage />);

        fireEvent.click(rendered.getAllByText('Для дома')[0]);
        fireEvent.click(rendered.getAllByText('Для дома')[0]);
        expect(rendered.getByText('Iphone')).toBeInTheDocument();
        expect(rendered.getByText('lamp')).toBeInTheDocument();
    });

    it('should render ProductCard for electronics', () => {
        const rendered = render(<MainPage />);

        expect(rendered.getByText('lamp')).toBeInTheDocument();
        expect(rendered.getByText('lamp_desc')).toBeInTheDocument();
        expect(rendered.getByText('150 $')).toBeInTheDocument();
        expect(rendered.getAllByText('Для дома')[0]).toBeInTheDocument();
        expect(rendered.getByAltText('lamp')).toBeInTheDocument();
    });

    it('should render ProductCard for electronics', () => {
        const rendered = render(<MainPage />);

         fireEvent.click(rendered.getAllByText('Электроника')[0]);

        expect(rendered.getByText('Iphone')).toBeInTheDocument();
        expect(rendered.getByText('Iphone_desc')).toBeInTheDocument();
        expect(rendered.getByText('999 $')).toBeInTheDocument();
        expect(rendered.getAllByText('Электроника')[0]).toBeInTheDocument();
        expect(rendered.getByAltText('Iphone')).toBeInTheDocument();
    });

});