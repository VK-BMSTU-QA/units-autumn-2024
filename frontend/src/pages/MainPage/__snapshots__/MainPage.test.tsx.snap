// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test main page component shoulds contain only 1 product with category Electronics 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3 />
    <div
      class="main-page__parameters"
    />
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Настольная лампа
        </h2>
        <div
          class="product-card__description"
        >
          Говорят, что ее использовали в pixar
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Для дома
        </div>
      </div>
      <img
        alt="Настольная лампа"
        class="product-card__image"
        src="/lamp.png"
      />
    </div>
  </div>
</DocumentFragment>
`;
